src/SourcemapToolkit.SourcemapParser/SourceMap.cs

```diff
@@ -1,51 +1,133 @@
+        /// </summary>
+        public int Version;
+
+        /// <summary>
+        /// The name of the generated file to which this source map corresponds
+        /// </summary>
+        public string File;
+
+        /// <summary>
+        /// The raw, unparsed mappings entry of the soure map
+        /// </summary>
+        public string Mappings;
+
+        /// <summary>
+        /// The list of source files that were the inputs used to generate this output file
+        /// </summary>
+        public List<string> Sources;
+
+        /// <summary>
+        /// A list of known original names for entries in this file
+        /// </summary>
+        public List<string> Names;
+
+        /// <summary>
+        /// Parsed version of the mappings string that is used for getting original names and source positions
+        /// </summary>
+        public List<MappingEntry> ParsedMappings;
+
+        /// <summary>
+        /// Applies the mappings of a sub source map to the current source map
+        /// Each mapping to the supplied source file is rewritten using the supplied source map
+        /// <param name="submap">The submap to apply</param>
+        /// <param name="aSourceFile">The filename of the source file. If not specified, submap's File property will be used</param>
+        /// <returns>A new source map</returns>
+        /// </summary>
+        public SourceMap ApplySourceMap(SourceMap submap, string aSourceFile = null)
+        {
+            if (submap == null)
+            {
+                throw new ArgumentNullException(nameof(submap));
+            }
+
+            string sourceFile = aSourceFile;
+            if (sourceFile == null)
+            {
+                if (submap.File == null)
+                {
+                    throw new Exception("ApplySourceMap expects either the explicit source file to the map, or submap's 'file' property");
+                }
+
+                sourceFile = submap.File;
+            }
+
+            List<string> newSources = new List<string>();
+            List<string> newNames = new List<string>();
+            List<MappingEntry> newMappingEntries = new List<MappingEntry>();
+
+            SourceMap newSourceMap = new SourceMap
+            {
+                File = this.File,
+                Version = this.Version,
+                Sources = newSources,
+                Names = newNames,
+                ParsedMappings = newMappingEntries
+            };
+
+            // transform mappings in this source map
+            foreach (MappingEntry mappingEntry in this.ParsedMappings)
+            {
+                MappingEntry newMappingEntry = mappingEntry.Clone() as MappingEntry;
+                SourcePosition original = mappingEntry.OriginalSourcePosition;
+
+                if (mappingEntry.OriginalFileName == sourceFile && original != null)
+                {
+                    MappingEntry correspondingEntry = submap.GetMappingEntryForGeneratedSourcePosition(original);
```

Comment:
Don't you want the generated position here instead of the original? If I understand correctly, the scenario here is a.js becomes b.js which becomes c.js and generates ab.sourcemap.js and bc.sourcemap.js to generate ac.sourcemap.js. So you'd call apply sourcemap on ab.sourcemap.js with bc.sourcemap.js, right?